File: macos/Sources/Updater.swift
import AppKit
import Foundation

let updateRepo = "Jaybee4real/banshell"
let macInstallerAsset = "Banshell-macOS-Installer.pkg"

struct ReleaseInfo {
    let version: String
    let assetURL: URL
    let notes: String
}

enum UpdateOutcome {
    case upToDate
    case available(ReleaseInfo)
    case failed(String)
}

func parseVersion(_ text: String) -> [Int] {
    text.trimmingCharacters(in: CharacterSet(charactersIn: "v "))
        .split(separator: ".")
        .map { Int($0.prefix(while: { $0.isNumber })) ?? 0 }
}

func versionIsNewer(_ candidate: String, than current: String) -> Bool {
    let candidateParts = parseVersion(candidate)
    let currentParts = parseVersion(current)
    for index in 0..<max(candidateParts.count, currentParts.count) {
        let left = index < candidateParts.count ? candidateParts[index] : 0
        let right = index < currentParts.count ? currentParts[index] : 0
        if left != right { return left > right }
    }
    return false
}

final class Updater {
    static let shared = Updater()
    private var checking = false
    private var downloadingVersion: String?
    private var readyRelease: ReleaseInfo?
    private var readyInstaller: (url: URL, version: String)?
    var onStatusChange: (() -> Void)?

    var statusText: String? {
        if let readyInstaller { return "Install v\(readyInstaller.version) & Restart" }
        if let readyRelease { return "Download & Install v\(readyRelease.version)" }
        if let downloadingVersion { return "Downloading v\(downloadingVersion)…" }
        return nil
    }

    func actOnStatus() {
        if let readyInstaller {
            runInstaller(at: readyInstaller.url, version: readyInstaller.version)
        } else if let release = readyRelease {
            readyRelease = nil
            downloadSilent(release, autoInstall: true)
        }
    }

    private func notifyStatus() {
        DispatchQueue.main.async { self.onStatusChange?() }
    }

    func fetchLatest() -> UpdateOutcome {
        guard let url = URL(string: "https://api.github.com/repos/\(updateRepo)/releases/latest") else {
            return .failed("bad url")
        }
        var request = URLRequest(url: url)
        request.setValue("BANSHELL-Updater", forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 15
        var outcome: UpdateOutcome = .failed("no response")
        let semaphore = DispatchSemaphore(value: 0)
        URLSession.shared.dataTask(with: request) { data, _, error in
            defer { semaphore.signal() }
            if let error {
                outcome = .failed(error.localizedDescription)
                return
            }
            guard let data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tag = json["tag_name"] as? String else {
                outcome = .failed("malformed response")
                return
            }
            let notes = json["body"] as? String ?? ""
            let assets = json["assets"] as? [[String: Any]] ?? []
            guard let asset = assets.first(where: { ($0["name"] as? String) == macInstallerAsset }),
                  let urlString = asset["browser_download_url"] as? String,
                  let assetURL = URL(string: urlString) else {
                outcome = .failed("no macOS installer attached to the latest release")
                return
            }
            let version = tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
            if versionIsNewer(version, than: banshellVersion) {
                outcome = .available(ReleaseInfo(version: version, assetURL: assetURL, notes: notes))
            } else {
                outcome = .upToDate
            }
        }.resume()
        semaphore.wait()
        return outcome
    }

    func checkForUpdates(silent: Bool) {
        guard !checking else { return }
        checking = true
        DispatchQueue.global(qos: .utility).async {
            let outcome = self.fetchLatest()
            DispatchQueue.main.async {
                self.checking = false
                switch outcome {
                case .upToDate:
                    if !silent { self.showInfo("You're up to date", "BANSHELL v\(banshellVersion) is the latest version.") }
                case .available(let info):
                    if silent { self.handleSilentAvailable(info) } else { self.promptUpdate(info) }
                case .failed(let message):
                    if silent {
                        logLine("update check failed: \(message)")
                    } else {
                        self.showInfo("Couldn't check for updates", message)
                    }
                }
            }
        }
    }

    private func promptUpdate(_ info: ReleaseInfo) {
        let alert = NSAlert()
        alert.messageText = "Update available: v\(info.version)"
        var body = "You have v\(banshellVersion). Install v\(info.version) now? BANSHELL will briefly restart."
        let trimmedNotes = info.notes.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedNotes.isEmpty {
            body += "\n\nWhat's new:\n" + String(trimmedNotes.prefix(500))
        }
        alert.informativeText = body
        alert.addButton(withTitle: "Install & Restart")
        alert.addButton(withTitle: "Later")
        NSApp.activate(ignoringOtherApps: true)
        guard alert.runModal() == .alertFirstButtonReturn else { return }
        downloadSilent(info, autoInstall: true)
    }

    private func handleSilentAvailable(_ info: ReleaseInfo) {
        let config = loadConfig() ?? Config.defaults
        guard config.autoDownloadOn else {
            readyRelease = info
            notifyStatus()
            logLine("update v\(info.version) available — download off, waiting for user")
            return
        }
        downloadSilent(info, autoInstall: config.autoInstallOn)
    }

    private func downloadSilent(_ info: ReleaseInfo, autoInstall: Bool) {
        downloadingVersion = info.version
        readyRelease = nil
        readyInstaller = nil
        notifyStatus()
        DispatchQueue.global(qos: .userInitiated).async {
            let destination = FileManager.default.temporaryDirectory
                .appendingPathComponent("Banshell-\(info.version).pkg")
            try? FileManager.default.removeItem(at: destination)
            var downloadError: String?
            let semaphore = DispatchSemaphore(value: 0)
            URLSession.shared.downloadTask(with: info.assetURL) { tempURL, _, error in
                defer { semaphore.signal() }
                if let error { downloadError = error.localizedDescription; return }
                guard let tempURL else { downloadError = "download produced no file"; return }
                do {
                    try FileManager.default.moveItem(at: tempURL, to: destination)
                } catch {
                    downloadError = error.localizedDescription
                }
            }.resume()
            semaphore.wait()

            DispatchQueue.main.async {
                self.downloadingVersion = nil
                if let downloadError {
                    logLine("update download failed: \(downloadError)")
                    self.notifyStatus()
                    return
                }
                if autoInstall {
                    self.runInstaller(at: destination, version: info.version)
                } else {
                    self.readyInstaller = (destination, info.version)
                    logLine("update v\(info.version) downloaded — install off, waiting for user")
                    self.notifyStatus()
                }
            }
        }
    }

    private func runInstaller(at path: URL, version: String) {
        let shell = "/usr/sbin/installer -pkg '\(path.path)' -target /"
        let source = "do shell script \"\(appleScriptEscaped(shell))\" with administrator privileges"
        logLine("installing update v\(version)")
        DispatchQueue.global(qos: .userInitiated).async {
            var errorInfo: NSDictionary?
            NSAppleScript(source: source)?.executeAndReturnError(&errorInfo)
            if let errorInfo {
                let code = errorInfo[NSAppleScript.errorNumber] as? Int ?? 0
                if code != -128 {
                    let message = errorInfo[NSAppleScript.errorMessage] as? String ?? "install failed"
                    DispatchQueue.main.async { self.showInfo("Update failed", message) }
                }
            }
        }
    }

    private func showInfo(_ title: String, _ message: String) {
        let alert = NSAlert()
        alert.messageText = title
        alert.informativeText = message
        NSApp.activate(ignoringOtherApps: true)
        alert.runModal()
    }
}

File: windows/src/Updater.cs
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Reflection;
using System.Text.Json;

namespace Banshell;

public static class Updater
{
    private const string Repo = "Jaybee4real/banshell";
    private const string InstallerAsset = "Banshell-Setup.exe";
    private static bool checking;

    public static string CurrentVersion
    {
        get
        {
            var version = Assembly.GetExecutingAssembly().GetName().Version;
            return version == null ? "0.0.0" : $"{version.Major}.{version.Minor}.{version.Build}";
        }
    }

    public static int[] ParseVersion(string text)
    {
        return text.TrimStart('v', ' ').Split('.')
            .Select(part => int.TryParse(new string(part.TakeWhile(char.IsDigit).ToArray()), out var value) ? value : 0)
            .ToArray();
    }

    public static bool IsNewer(string candidate, string current)
    {
        var left = ParseVersion(candidate);
        var right = ParseVersion(current);
        for (int index = 0; index < Math.Max(left.Length, right.Length); index++)
        {
            int a = index < left.Length ? left[index] : 0;
            int b = index < right.Length ? right[index] : 0;
            if (a != b) return a > b;
        }
        return false;
    }

    private record ReleaseInfo(string Version, string AssetUrl, string Notes);

    private static async Task<ReleaseInfo?> FetchLatestAsync()
    {
        using var client = new HttpClient();
        client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("BANSHELL-Updater", "1.0"));
        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
        client.Timeout = TimeSpan.FromSeconds(15);
        var json = await client.GetStringAsync($"https://api.github.com/repos/{Repo}/releases/latest");
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        var tag = root.GetProperty("tag_name").GetString() ?? "";
        var version = tag.StartsWith('v') ? tag[1..] : tag;
        var notes = root.TryGetProperty("body", out var bodyElement) ? bodyElement.GetString() ?? "" : "";
        if (!root.TryGetProperty("assets", out var assets)) return null;
        foreach (var asset in assets.EnumerateArray())
        {
            if (asset.GetProperty("name").GetString() == InstallerAsset)
                return new ReleaseInfo(version, asset.GetProperty("browser_download_url").GetString() ?? "", notes);
        }
        return null;
    }

    public static async Task CheckAsync(bool silent)
    {
        if (checking) return;
        checking = true;
        try
        {
            ReleaseInfo? info;
            try
            {
                info = await FetchLatestAsync();
            }
            catch (Exception error)
            {
                if (!silent) MessageBox.Show($"Couldn't check for updates: {error.Message}", "BANSHELL");
                return;
            }
            if (info == null)
            {
                if (!silent) MessageBox.Show("No Windows installer found in the latest release.", "BANSHELL");
                return;
            }
            if (!IsNewer(info.Version, CurrentVersion))
            {
                if (!silent) MessageBox.Show($"BANSHELL is up to date (v{CurrentVersion}).", "BANSHELL");
                return;
            }
            var notesPreview = string.IsNullOrWhiteSpace(info.Notes) ? "" : "\n\nWhat's new:\n" + Truncate(info.Notes, 500);
            var choice = MessageBox.Show(
                $"Update available: v{CurrentVersion} → v{info.Version}\n\nDownload and install now? BANSHELL will restart.{notesPreview}",
                "BANSHELL Update", MessageBoxButtons.YesNo, MessageBoxIcon.Information);
            if (choice == DialogResult.Yes) await DownloadAndRunAsync(info);
        }
        finally
        {
            checking = false;
        }
    }

    private static async Task DownloadAndRunAsync(ReleaseInfo info)
    {
        var target = Path.Combine(Path.GetTempPath(), $"Banshell-Setup-{info.Version}.exe");
        try
        {
            using (var client = new HttpClient())
            {
                client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("BANSHELL-Updater", "1.0"));
                client.Timeout = TimeSpan.FromMinutes(5);
                var bytes = await client.GetByteArrayAsync(info.AssetUrl);
                await File.WriteAllBytesAsync(target, bytes);
            }
        }
        catch (Exception error)
        {
            MessageBox.Show($"Update download failed: {error.Message}", "BANSHELL");
            return;
        }
        Process.Start(new ProcessStartInfo(target, "/SILENT /CLOSEAPPLICATIONS /RESTARTAPPLICATIONS")
        {
            UseShellExecute = true,
        });
        Application.Exit();
    }

    private static string Truncate(string text, int max) =>
        text.Length <= max ? text : text[..max];
}
